ENIGMAsystem/SHELL/Universal_System/Resources/sprites.cpp

```diff
@@ -0,0 +1,258 @@
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include "sprites_internal.h"
+#include "Universal_System/image_formats.h"
+#include "Universal_System/nlpo2.h"
+#include "Graphics_Systems/graphics_mandatory.h"
+#include "Widget_Systems/widgets_mandatory.h"
+
+#include <string>
+#include <iostream>
+
+using enigma::Sprite;
+using enigma::sprites;
+using enigma::nlpo2dc;
+using enigma::TexRect;
+using enigma::Color;
+using enigma::RawImage;
+
+namespace {
+
+Sprite sprite_add_helper(std::string filename, int imgnumb, bool precise, bool transparent, bool smooth, bool preload, int x_offset, int y_offset, bool mipmap) {
+  unsigned int width, height, fullwidth, fullheight;
+  unsigned char *pxdata = enigma::image_load(
+      filename, &width, &height, &fullwidth, &fullheight, &imgnumb, false);
+  
+  unsigned cellwidth = width / imgnumb;
+  Sprite ns(cellwidth, height, x_offset, y_offset);
+  ns.SetBBox(0, 0, cellwidth, height);
+
+  if (pxdata == NULL) {
+    DEBUG_MESSAGE("ERROR - Failed to append sprite to index!", MESSAGE_TYPE::M_ERROR);
+    return ns;
+  }
+
+  // If sprite transparent, set the alpha to zero for pixels that should be
+  // transparent from lower left pixel color
+  if (transparent) {
+    Color c = enigma::image_get_pixel_color(pxdata, width, height, 0, height - 1);
+    enigma::image_swap_color(pxdata, width, height, c, Color {0, 0, 0, 0});
+  }
+  
+  std::vector<RawImage> rawSubimages = enigma::image_split(pxdata, width, height, imgnumb);
+  for (const RawImage& i : rawSubimages) {
+    ns.AddSubimage(i.pxdata, i.w, i.h, ((precise) ? enigma::ct_precise : enigma::ct_bbox), i.pxdata, mipmap);
+  }
+  
+  delete[] pxdata;
+  
+  return ns;
+}
+
+}
+
+namespace enigma_user {
+
+int sprite_get_width(int sprid) {
+  return sprites.get(sprid).width;
+}
+
+int sprite_get_height(int sprid) {
+  return sprites.get(sprid).height;
+}
+
+gs_scalar sprite_get_texture_width_factor(int sprid, int subimg) {
+  return sprites.get(sprid).GetTextureRect(subimg).w;
+}
+
+gs_scalar sprite_get_texture_height_factor(int sprid, int subimg) {
+  return sprites.get(sprid).GetTextureRect(subimg).h;
+}
+
+int sprite_get_bbox_bottom(int sprid) {
+  return sprites.get(sprid).bbox.y2();
+}
+
+int sprite_get_bbox_left(int sprid) {
+  return sprites.get(sprid).bbox.x;
+}
+
+int sprite_get_bbox_right(int sprid) {
+  return sprites.get(sprid).bbox.x2();
+}
+
+int sprite_get_bbox_top(int sprid) {
+  return sprites.get(sprid).bbox.y;
+}
+
+int sprite_get_bbox_mode(int sprid) {
+  return sprites.get(sprid).bbox_mode;
+}
+
+int sprite_get_bbox_bottom_relative(int sprid) {
+  const Sprite& spr = sprites.get(sprid);
+  return spr.bbox.y2() - spr.yoffset;
+}
+
+int sprite_get_bbox_left_relative(int sprid) {
+  const Sprite& spr = sprites.get(sprid);
+  return spr.bbox.x - spr.xoffset;
+}
+
+int sprite_get_bbox_right_relative(int sprid) {
+  const Sprite& spr = sprites.get(sprid);
+  return spr.bbox.x2() - spr.xoffset;
```

Comment:
I think this is supposed to be width - xoffset. Worth checking.

Reply:
should be correct as all tests are passing